### ISSUE TITLE
Authentication Identity Says no Email Service Configured While it Uses the Email Service Configured

### ISSUE THREAD

**Issue by DanJ210** [dotnet/aspnetcore#18730]
### Describe the bug
I've followed the steps in the documentation to add authentication. Tried doing it to an app I already have built but failed and the docs only explain how to add authentication on a new one. So I created a new one, exactly as described in the docs. Email service has been hit or miss.

Email service was working, intermittently meaning one day it worked once by sending out an email then stopped sending out email when there were no code changes, next day it would send out no email though the email service code was being called and there were no code changes. Day after that it was back to sending out email again, still no code changes. Hoped all would be fine but now if I try to register a user as described in the docs, step by step, I'm told that the app has no real email sender registered so I can click the link to confirm which is what you experience before you add the email service as directed in docs. Yet the code within `EmailSender.cs` is hit, it succeeds and there's no error or exception. In fact the screenshot below shows the email activity from the SendGrid dashboard as sent and processed. The email is actually never received though because, another issue, sending to Hotmail doesn't work, the email doesn't show up in any folder, no junk, no nothing. Well I received the first email I ever sent one time, since then Hotmail has received no further emails from SendGrid, Gmail has no problem at all and has received all that were processed and sent. So the code does work yet the Identity library seems to have lost touch with itself and thinks there's no email service when it's actually using it. 

I'll include a link to the repo I just put on GitHub so the entire codebase I'm using can be viewed. I haven't modified the code or steps within the documentation at all. Just been spending days testing emails because it seems to work then not work. And then Hotmail accounts don't receive email at all from SendGrid which is the service that the documentation suggests that we use. I could just create another app since I've not made any changes of my own but if I'm getting issues this heavily on boilerplate then I don't think another copy would be a solution.

![SendGrid Activity - Email Sent to Hotmail Account](https://user-images.githubusercontent.com/16159685/73597067-4a744980-44f6-11ea-811b-c020ec567c80.png)

### To Reproduce
- Account confirmation and password recovery in ASP.NET Core
[Authentication Asp.Net Core 3.1 Documentation](https://docs.microsoft.com/en-us/aspnet/core/security/authentication/accconfirm?view=aspnetcore-3.1&tabs=visual-studio)
Haven't gone past test password reset emails working because the confirm account email section is working half of the time and now the app thinks there isn't one, below is the stopping point in the docs.
[Register, confirm email, and reset password section](https://docs.microsoft.com/en-us/aspnet/core/security/authentication/accconfirm?view=aspnetcore-3.1&tabs=visual-studio#register-confirm-email-and-reset-password)

[GitHub Repo](https://github.com/DanJ210/LoginAccounts)

Steps to reproduce for me were following the documentation with the links provided right above and registering users to go through the confirm email flow which has worked before. I made no code changes or additions so I couldn't described what could have happened to cause it not to detect the email service while it uses the email service and sends out an email that Hotmail no longer receives.

### Further technical details
- ASP.NET Core version - 3.1
- Include the output of `dotnet --info`
.NET Core SDK (reflecting any global.json):
 Version:   3.1.101
 Commit:    b377529961

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.18363
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\3.1.101\

Host (useful for support):
  Version: 3.1.1
  Commit:  a1388f194c

.NET Core SDKs installed:
  2.1.509 [C:\Program Files\dotnet\sdk]
  2.1.802 [C:\Program Files\dotnet\sdk]
  2.2.402 [C:\Program Files\dotnet\sdk]
  3.0.100 [C:\Program Files\dotnet\sdk]
  3.1.100 [C:\Program Files\dotnet\sdk]
  3.1.101 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.13 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.15 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.13 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.15 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.13 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.15 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.1 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download
- The IDE (VS / VS Code/ VS4Mac) you're running on, and it's version
VS Code 1.41.1 (User Setup)
and
Visual Studio 2019 16.4.3

**Comment by DanJ210** [comment]
Another relevant finding...

Creating a brand new app to start from scratch again following the documentation, basically the same steps as above, same app as above.

<details><summary>Steps in documentation where break lies</summary>
- [Starting point](https://docs.microsoft.com/en-us/aspnet/core/security/authentication/accconfirm?view=aspnetcore-3.1&tabs=visual-studio#create-and-test-a-web-app-with-authentication)
- [Code working up to this point]([Register, confirm and reset email](https://docs.microsoft.com/en-us/aspnet/core/security/authentication/accconfirm?view=aspnetcore-3.1&tabs=visual-studio#register-confirm-email-and-reset-password)
- [Adding external accounts to pipeline doesn't use email service](https://docs.microsoft.com/en-us/aspnet/core/security/authentication/social/?view=aspnetcore-3.1&tabs=visual-studio)
</details>

Following the docs to [Register, confirm and reset email](https://docs.microsoft.com/en-us/aspnet/core/security/authentication/accconfirm?view=aspnetcore-3.1&tabs=visual-studio#register-confirm-email-and-reset-password) and testing results in email sending, code is working. I'll skip the next short section which is adjusting default expiration tokens and inactivity timeouts etc... Defaults are fine for now there should be no problem with skipping this part. The steps after is what doesn't work.

[Combining social and local login accounts](https://docs.microsoft.com/en-us/aspnet/core/security/authentication/accconfirm?view=aspnetcore-3.1&tabs=visual-studio#combine-social-and-local-login-accounts) shows the next step is to first [add external login accounts](https://docs.microsoft.com/en-us/aspnet/core/security/authentication/social/?view=aspnetcore-3.1&tabs=visual-studio). I chose to add both Microsoft and Facebook but you could add only one of either and the same results happen. The code below is all that's needed to be added to the `Startup.cs` file. Simple and easy. But testing results in the email service not being used when registering with these external accounts. Setting breakpoints in `EmailSender.cs` as suggested by the docs shows that when you register a plain local account, the code is hit and everything is all good. When you try and register with either Facebook or Microsoft, the `EmailSender.cs` constructor is hit as usual, but then no other actual methods are called. It would seem that there's something within the pipeline that's failing to use the email service when using the external services to register. The `EmailSender.cs` constructor is always hit twice, once before its inner code is called upon and once after everything is done. This continues to happen in both cases except with external accounts the constructor is simply called twice and no other body code is ever called like when registering local accounts.

```
            .AddMicrosoftAccount(microsoftOptions => {
                microsoftOptions.ClientId = Configuration["Authentication:Microsoft:ClientId"];
                microsoftOptions.ClientSecret = Configuration["Authentication:Microsoft:ClientSecret"];
            })
            .AddFacebook(facebookOptions => {
                facebookOptions.AppId = Configuration["Authentication:Facebook:ClientId"];
                facebookOptions.AppSecret = Configuration["Authentication:Facebook:ClientSecret"];
            });
```

This is the same development point as the app shared in the repo above. Both are exact copies of each other. Except the one above is throwing a different issue altogether along with both not utilizing the email service when using external accounts. So there's two issues except the second app is reproducing one of them for now but this is also what the first app did. I'm trying to give as much info as possible to help as much as possible because I'd really like to use these authentication schemes for an app I've needed to start creating a couple weekends ago except haven't gotten past this very early stage. Hopefully not a sign of my luck for things to come lol.

**Comment by DanJ210** [comment]
Let me know if I've been too confusing or if I can add anything else. I'm anxious to find out if there's something I'm doing completely wrong or if someone else is able to replicate the issue.

**Comment by HaoK** [comment]
So just to clarify, you don't mean its half working, you just mean password based registration works, and social logins do not currently send account confirmation correct?

If that's the case, yes that's an issue we have fixed for 3.1.2 and 5.0, see https://github.com/dotnet/aspnetcore/pull/18206

You can scaffold out the external login page and move the send email code similar to the fix in the mean time to work around the issue.

**Comment by DanJ210** [comment]
@HaoK I had a lot of working then not working at first but so far what's consistent is local account registration works and calls upon the right email service channels. Social registrations call upon the email service but then never actually use it. Meaning the constructor is called and options/settings are setup but none of the further methods to generate and send an email are called, when using social registrations. If that's solved then great.

There's the first main issue I posted about which is the reason I uploaded the project to GitHub. When registering a local account the app states that there's no email service configured while it's using the email service to send out email. I've detailed it much further in the OP.

**Comment by blowdart** [comment]
Until the patch comes out in 3.1.2 and you upgrade we can't go much further here. It's supposed to address the issue we believe you are describing. 

Once 3.1.2 drops please upgrade and follow the scaffolding instructions if the default behaviour is not as you expect.

**Comment by DanJ210** [comment]
@blowdart I'm ok with that. Are you able to easily share with me a page I can follow which has a timeline for version releases? No worries if not.

Also, I believe the "no email provider configured" issue might be a more rare one which happens. If there's not another ticket about it then it might be something that hasn't been addressed. I'll update this post if the update drops within a quick enough time frame before this gets closed out.

**Comment by blowdart** [comment]
Huh, I don't know of a timeline page. @anurse do we have one?

**Comment by anurse** [comment]
You can look at the [.NET Core Roadmap](https://github.com/dotnet/core/blob/master/roadmap.md), which covers major/minor releases. We have no committed schedule for patch releases, but we try to get them out frequently (around monthly). For 3.1.2, we're expecting a release next week (but no guarantees ;)).

**Comment by HaoK** [comment]
Closing this out because this should be fixed in 3.1.2 and 5.0, feel free to reopen or file a new issue if you still see any issues in the new versions that should have the fix @DanJ210
